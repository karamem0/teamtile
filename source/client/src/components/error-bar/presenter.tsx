//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { WarningIcon } from '@fluentui/react-icons-mdl2';
import { Alert } from '@fluentui/react-northstar';

import { css } from '@emotion/react';

interface ErrorBarProps {
  message: string | null | undefined,
  onVisibleChange: () => void
}

export default React.memo(function ErrorBar ({
  message,
  onVisibleChange
}: ErrorBarProps): React.ReactElement | null {

  return (
    <div
      css={css`
        margin-bottom: 1rem;
      `}>
      <Alert
        content={message}
        dismissible
        icon={
          <WarningIcon />
        }
        variables={{ urgent: true }}
        onVisibleChange={onVisibleChange} />
    </div>
  );

});
