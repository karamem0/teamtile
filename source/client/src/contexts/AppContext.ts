//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

import React from 'react';
import { Client } from '@microsoft/microsoft-graph-client';

const AppContext = React.createContext<[
  client?: Client,
  setClient?: React.Dispatch<React.SetStateAction<Client | undefined>>,
  error?: string,
  setError?: React.Dispatch<React.SetStateAction<string | undefined>>
]>([]);

export default AppContext;
