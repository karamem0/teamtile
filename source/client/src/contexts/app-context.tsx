//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

import React from 'react';
import { Client } from '@microsoft/microsoft-graph-client';

interface AppContextValue {
  client?: Client,
  setClient?: React.Dispatch<React.SetStateAction<Client | undefined>>,
  error?: string,
  setError?: React.Dispatch<React.SetStateAction<string | undefined>>
}

const AppContext = React.createContext<AppContextValue>({});

interface AppContextProviderProps {
  value: AppContextValue,
  children: React.ReactNode | undefined
}

const AppContextProvider = ({ children, value }: AppContextProviderProps): React.ReactElement => {

  return (
    <AppContext.Provider value={value}>
      {children}
    </AppContext.Provider>
  );

};

interface AppContextConsumerProps {
  children: (value: AppContextValue) => React.ReactNode
}

const AppContextConsumer = ({ children }: AppContextConsumerProps): React.ReactElement => {

  return (
    <AppContext.Consumer>
      {children}
    </AppContext.Consumer>
  );

};

const useAppContext = (): AppContextValue => React.useContext(AppContext);

export {
  AppContextProvider,
  AppContextConsumer,
  useAppContext
};
