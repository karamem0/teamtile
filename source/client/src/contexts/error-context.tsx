//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

// React
import React from 'react';

interface ErrorContextValue {
  error?: string,
  setError?: React.Dispatch<React.SetStateAction<string | undefined>>
}

const ErrorContext = React.createContext<ErrorContextValue>({});

interface ErrorContextProviderProps {
  children?: React.ReactNode
}

export const ErrorContextProvider = ({ children }: ErrorContextProviderProps): React.ReactElement | null => {

  const [ error, setError ] = React.useState<string>();

  return (
    <ErrorContext.Provider
      value={{
        error,
        setError
      }}>
      {children}
    </ErrorContext.Provider>
  );

};

interface ErrorContextConsumerProps {
  children: (value: ErrorContextValue) => React.ReactNode
}

export const ErrorContextConsumer = ({ children }: ErrorContextConsumerProps): React.ReactElement | null => {

  return (
    <ErrorContext.Consumer>
      {children}
    </ErrorContext.Consumer>
  );

};

export const useErrorContext = (): ErrorContextValue => React.useContext(ErrorContext);
