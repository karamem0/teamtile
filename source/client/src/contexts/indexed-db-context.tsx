//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

import React from 'react';
import Dexie from 'dexie';

interface IndexedDbValue {
  db?: Dexie
}

const IndexedDb = React.createContext<IndexedDbValue>({});

interface IndexedDbProviderProps {
  children: React.ReactNode
}

const IndexedDbProvider = ({ children }: IndexedDbProviderProps): React.ReactElement => {

  const [ dexie, setDexie ] = React.useState<Dexie>();

  React.useEffect(() => {
    const dexie = new Dexie('teamtile');
    dexie.version(2).stores({
      teams:
        '&id, internalId, name, description, visibility, url,' +
        ' icon.data,' +
        ' channel.count, channel.nextLink,' +
        ' channel.values.id, channel.values.name, channel.values.url,' +
        ' member.count, member.nextLink,' +
        ' member.values.id, member.values.name, member.values.url,' +
        ' drive.id, drive.url,' +
        ' timestamp'
    });
    setDexie(dexie);
  }, []);

  return (
    <IndexedDb.Provider value={{ db: dexie }}>
      {children}
    </IndexedDb.Provider>
  );

};

interface IndexedDbConsumerProps {
  children: (value: IndexedDbValue) => React.ReactNode
}

const IndexedDbConsumer = ({ children }: IndexedDbConsumerProps): React.ReactElement => {

  return (
    <IndexedDb.Consumer>
      {children}
    </IndexedDb.Consumer>
  );

};

const useIndexedDb = (): IndexedDbValue => React.useContext(IndexedDb);

export {
  IndexedDbProvider,
  IndexedDbConsumer,
  useIndexedDb
};
