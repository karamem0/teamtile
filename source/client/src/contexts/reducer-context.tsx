//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import {
  setChannels,
  setDrives,
  setFilter,
  setKeys,
  setLoading,
  setMemberIcons,
  setMembers,
  setTeamIcons,
  setTeams
} from '../reducers/action';
import { reducer } from '../reducers/reducer';
import { KeyValue } from '../types/common';
import {
  Channel,
  Drive,
  Member,
  Team
} from '../types/entity';
import { DispatchAction } from '../types/reducer';
import {
  ItemKey,
  Loading,
  State
} from '../types/state';

interface ReducerContextValue {
  state: State,
  dispatchers: {
    dispatchChannels: DispatchAction<Map<ItemKey, Channel[]>>,
    dispatchDrives: DispatchAction<Map<ItemKey, Drive>>,
    dispatchFilter: DispatchAction<string | null>,
    dispatchKeys: DispatchAction<ItemKey[]>,
    dispatchLoading: DispatchAction<Loading>,
    dispatchMemberIcons: DispatchAction<KeyValue<ItemKey, Map<string, string | null>>>,
    dispatchMembers: DispatchAction<Map<ItemKey, Member[]>>,
    dispatchTeamIcons: DispatchAction<Map<ItemKey, string | null>>,
    dispatchTeams: DispatchAction<Map<ItemKey, Team>>
  }
}

const ReducerContext = React.createContext<ReducerContextValue | null>(null);

interface ReducerContextProviderProps {
  children: React.ReactNode
}

const initialState = {
  loading: Loading.none,
  itemFilter: null,
  items: []
};

export const ReducerContextProvider = ({ children }: ReducerContextProviderProps): React.ReactElement | null => {

  const [ state, dispatch ] = React.useReducer(reducer, initialState);

  const dispatchers = React.useMemo(() => ({
    dispatchChannels: (payload: Map<ItemKey, Channel[]>) => dispatch(setChannels(payload)),
    dispatchDrives: (payload: Map<string, Drive>) => dispatch(setDrives(payload)),
    dispatchFilter: (payload: string | null) => dispatch(setFilter(payload)),
    dispatchKeys: (payload: ItemKey[]) => dispatch(setKeys(payload)),
    dispatchLoading: (payload: Loading) => dispatch(setLoading(payload)),
    dispatchMemberIcons: (payload: KeyValue<ItemKey, Map<string, string | null>>) => dispatch(setMemberIcons(payload)),
    dispatchMembers: (payload: Map<ItemKey, Member[]>) => dispatch(setMembers(payload)),
    dispatchTeamIcons: (payload: Map<string, string | null>) => dispatch(setTeamIcons(payload)),
    dispatchTeams: (payload: Map<string, Team>) => dispatch(setTeams(payload))
  }), []);

  return (
    <ReducerContext.Provider
      value={{
        state,
        dispatchers
      }}>
      {children}
    </ReducerContext.Provider>
  );

};

export const useReducerContext = (): ReducerContextValue => {
  const value = React.useContext(ReducerContext);
  if (!value) {
    throw new Error('The context is not initialzed: ReducerContext');
  }
  return value;
};
