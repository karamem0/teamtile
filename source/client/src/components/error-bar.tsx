//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { WarningIcon } from '@fluentui/react-icons-mdl2';
import { Alert } from '@fluentui/react-northstar';

import { css } from '@emotion/react';

import { useErrorContext } from '../contexts/error-context';

export const ErrorBar = (): React.ReactElement | null => {

  const { error, setError } = useErrorContext();

  return (
    <div
      css={css`
        margin-bottom: 1rem;
      `}>
      <Alert
        content={error}
        dismissible
        icon={
          <WarningIcon />
        }
        variables={{ urgent: true }}
        onVisibleChange={() => setError && setError(null)} />
    </div>
  );

};
