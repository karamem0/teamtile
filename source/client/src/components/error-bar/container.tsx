//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { useErrorContext } from '../../contexts/error-context';

import Presenter from './presenter';

export default function ErrorBar (): React.ReactElement | null {

  const { error, setError } = useErrorContext();

  const handleVisibleChange = React.useCallback(() => {
    setError(null);
  }, [
    setError
  ]);

  return (
    <Presenter
      message={error}
      onVisibleChange={handleVisibleChange} />
  );

}
