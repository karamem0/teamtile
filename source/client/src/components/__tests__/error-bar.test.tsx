//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { render, screen } from '@testing-library/react';

import * as ErrorContext from '../../contexts/error-context';
import { ErrorBar } from '../error-bar';

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

describe('ErrorBar', () => {

  it('create shapshot', () => {
    jest
      .spyOn(ErrorContext, 'useErrorContext')
      .mockReturnValue({
        error: 'something went wrong',
        setError: jest.fn()
      });
    render(<ErrorBar />);
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

});
