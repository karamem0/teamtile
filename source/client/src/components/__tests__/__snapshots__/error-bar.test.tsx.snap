// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBar create shapshot 1`] = `
.emotion-0 {
  margin-bottom: 1rem;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      <div
        class="ui-alert"
        data-uses-unhanded-props="true"
      >
        <div
          class="ui-box ui-alert__icon"
          data-uses-unhanded-props="true"
        >
          <span
            aria-hidden="true"
            class="root_3aeb045a"
          >
            <svg
              class="svg_3aeb045a"
              focusable="false"
              viewBox="0 0 2048 2048"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1920 1920H0L960 0l960 1920zM207 1792h1506L960 286 207 1792zm817-1024v640H896V768h128zm-128 768h128v128H896v-128z"
              />
            </svg>
          </span>
        </div>
        <div
          class="ui-box ui-alert__body"
          data-uses-unhanded-props="true"
          id="alert-body-1"
        >
          <div
            class="ui-box ui-alert__content"
            data-uses-unhanded-props="true"
            dir="auto"
          >
            something went wrong
          </div>
        </div>
        <button
          aria-describedby="alert-body-1"
          class="ui-alert__dismissaction"
          data-aa-class="Alert__dismissAction"
          data-uses-unhanded-props="true"
        >
          <span
            class="ui-box ui-alert__dismissaction__content"
            data-uses-unhanded-props="true"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
